src/request/swap-iframe/SwapIFrameApi.js

```diff
@@ -0,0 +1,385 @@
+            throw new Errors.InvalidRequestError('HTLC hashes do not match');
+        }
+
+        // TODO: Validate timeouts of the two contracts
+        // Currently not possible because the NIM timeout is a block height, while the BTC timeout is a timestamp.
+        // And since we cannot trust the local device time to be accurate, and we don't have a reference for NIM blocks
+        // and their timestamps, we cannot compare the two.
+        // When it becomes possible to compare (with Nimiq 2.0 Albatross), the redeem HTLC must have a later timeout
+        // than the funding HTLC.
+
+        /** @type {KeyguardRequest.SignSwapTransactionsResult} */
+        const result = {};
+
+        if (parsedRequest.fund.type === 'NIM' && storedRequest.fund.type === 'NIM') {
+            await loadNimiq();
+
+            const privateKey = new Nimiq.PrivateKey(Nimiq.BufferUtils.fromHex(privateKeys.nim));
+            const publicKey = Nimiq.PublicKey.derive(privateKey);
+
+            const transaction = Nimiq.Transaction.fromPlain({
+                ...storedRequest.fund.transaction.toPlain(),
+                data: parsedRequest.fund.htlcData,
+                // Must be the exact object reference, which gets lost with toPlain()
+                recipient: Nimiq.Address.CONTRACT_CREATION,
+            });
+
+            const signature = Nimiq.Signature.create(privateKey, publicKey, transaction.serializeContent());
+
+            /** @type {KeyguardRequest.SignatureResult} */
+            result.nim = {
+                publicKey: publicKey.serialize(),
+                signature: signature.serialize(),
+            };
+        }
+
+        if (parsedRequest.fund.type === 'BTC' && storedRequest.fund.type === 'BTC') {
+            const inputs = storedRequest.fund.inputs;
+
+            // Sort inputs by tx hash ASC, then index ASC
+            inputs.sort((a, b) => {
+                if (a.hash !== b.hash) return a.hash < b.hash ? -1 : 1;
+                return a.index - b.index;
+            });
+
+            // Construct outputs
+            const outputs = [{
+                address: parsedRequest.fund.htlcAddress,
+                value: storedRequest.fund.recipientOutput.value,
+            }];
+
+            // Add change output
+            if (storedRequest.fund.changeOutput) {
+                // The address is set in SignSwap after password entry.
+                outputs.push(/** @type {{address: string, value: number}} */ (storedRequest.fund.changeOutput));
+            }
+
+            // Sort outputs by value ASC, then address ASC
+            outputs.sort((a, b) => (a.value - b.value) || (a.address < b.address ? -1 : 1));
+
+            // Construct transaction
+            const psbt = new BitcoinJS.Psbt({ network: BitcoinUtils.Network });
+
+            // Add inputs
+            // @ts-ignore Argument of type 'Uint8Array' is not assignable to parameter of type 'Buffer'.
```

Comment:
No, because the browser-polyfilled "Buffer" (which is a Uint8Array internally) is still not the same as the _type_ `Buffer`.

Reply:
And making the type use `Buffer` instead, just shifts the TS errors from this here usage to the creation of the object, plus it would be using non-browser types.

Reply:
Maybe this would be cleaner though. I opened https://github.com/nimiq/keyguard/issues/432 to investigate this after this is merged.